diff --git a/init/main.c b/init/main.c
index c4778edae797..474385dd147a 100644
--- a/init/main.c
+++ b/init/main.c
@@ -943,9 +943,12 @@ void start_kernel(void)
 				  __stop___param - __start___param,
 				  -1, -1, NULL, &unknown_bootoption);
 	print_unknown_bootoptions();
-	if (!IS_ERR_OR_NULL(after_dashes))
+	if (!IS_ERR_OR_NULL(after_dashes)) {
+		/* elide the non-kernel params by puting a \0 just after the -- */
+		saved_command_line[after_dashes-static_command_line-1] = "\0";
 		parse_args("Setting init args", after_dashes, NULL, 0, -1, -1,
 			   NULL, set_init_arg);
+	}
 	if (extra_init_args)
 		parse_args("Setting extra init args", extra_init_args,
 			   NULL, 0, -1, -1, NULL, set_init_arg);
